### PR TITLE
feat(sdk): support batch for ppl

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description

1. (feature) support ppl batch
2. (fix) Solve the problem that the result of PPL returns more than one dimension under single data

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: example/speech_command/sc/evaluator.py [hunk 1]
```diff
@@ -65,9 +65,9 @@ def __init__(self) -> None:
 
     @torch.no_grad()
     def ppl(self, audio: Audio, **kw):
-        _tensor = self._pre(audio)
+        _tensor = self._pre([audio])
```
Comment (tianweidut) [inline example/speech_command/sc/evaluator.py:68]: If the users don't config batch size, the example code keeps the original structure?

File: example/mnist/mnist/evaluator.py [hunk 2]
```diff
@@ -16,13 +16,16 @@
...
     def __init__(self) -> None:
-        super().__init__()
+        super().__init__(ppl_batch_size=5)
         self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
         self.model = self._load_model(self.device)
 
     @api(Image(shape=(28, 28, 3)), JsonOutput())
-    def ppl(self, img: Image, **kw: t.Any) -> t.Tuple[t.List[int], t.List[float]]:  # type: ignore
-        data_tensor = self._pre(img)
+    def single_ppl(self, img: Image, **kw: t.Any) -> t.Tuple[int, t.List[float]]:  # type: ignore
```
Comment (tianweidut) [inline example/mnist/mnist/evaluator.py:24]: Cloud we remove the single_ppl function?  `@api` maybe handle batch input and output automatically

File: example/mnist/mnist/custom_evaluator.py [hunk 3]
```diff
@@ -166,19 +169,28 @@ def _load_model(self, device: torch.device) -> Net:
...
-            input.shape[0], input.shape[1]  # type: ignore
-        )
-        _image_array = PILImage.fromarray(_tensor.numpy())
-        _image = transforms.Compose(
-            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
-        )(_image_array)
-        return torch.stack([_image]).to(self.device)
+    def _pre(self, data: t.List[Image]) -> t.Any:
+        images = []
+        for i in data:
```
Comment (tianweidut) [inline example/mnist/mnist/custom_evaluator.py:174]: Can we use the torch array operation to replace the `for-item-append` operation?

File: client/starwhale/api/_impl/model.py [hunk 4]
```diff
@@ -223,57 +227,69 @@ def _starwhale_internal_run_ppl(self) -> None:
...
-        for ds_uri in self.context.dataset_uris:
-            _uri = URI(ds_uri, expected_type=URIType.DATASET)
-            consumption = get_dataset_consumption(
-                dataset_uri=_uri, session_id=self.context.version
-            )
-            loader = get_data_loader(_uri, session_consumption=consumption)
-
+        for uri_str in self.context.dataset_uris:
+            _uri = URI(uri_str, expected_type=URIType.DATASET)
+            ds = Dataset.dataset(_uri)
```
Comment (tianweidut) [inline client/starwhale/api/_impl/model.py:232]: Dataset.dataset function can handle uri string type automatically.

File: example/speech_command/sc/evaluator.py (hunk 1 above)
Comment (goldenxinxing) [reply example/speech_command/sc/evaluator.py:68]: the code need changed from 'def ppl(self, audio: Audio, **kw):' -> 'def ppl(self, audios: List[Audio], **kw):'

File: example/mnist/mnist/evaluator.py (hunk 2 above)
Comment (goldenxinxing) [reply example/mnist/mnist/evaluator.py:24]: this depends on service implement,  the currently implement seems to only support single, is it? @jialeicui

File: client/starwhale/api/_impl/model.py (hunk 4 above)
Comment (goldenxinxing) [reply client/starwhale/api/_impl/model.py:232]: this section also need URI's function : `index_with_dataset=[
                                f"{_uri.object}_#@#_{row.index}" for row in rows
                            ],`

File: client/starwhale/api/_impl/dataset/model.py [hunk 5]
```diff
@@ -259,6 +260,17 @@ def _get_data_loader(
 
         return _loader
 
+    def batch_iter(
+        self, batch_size: int = 1, drop_not_full: bool = False
+    ) -> t.Iterator[t.List[DataRow]]:
+        """Batch data into lists of length n. The last batch may be shorter."""
+        it = iter(self.__iter__())
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/api/_impl/dataset/model.py]: done

File: client/starwhale/api/_impl/model.py [hunk 6]
```diff
@@ -223,57 +227,69 @@ def _starwhale_internal_run_ppl(self) -> None:
...
-
-                result_storage.save(
-                    data_id=_idx_with_ds,
-                    index=_idx,
-                    result=_result,
-                    annotations={} if self.ignore_annotations else _annotations,
-                )
+                for (_idx, _data, _annotations), _result in zip(rows, _results):
+                    cnt += 1
+                    _idx_with_ds = f"{_uri.object}_#@#_{_idx}"
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply client/starwhale/api/_impl/model.py:285]: done

File: client/starwhale/api/_impl/model.py (hunk 4 above)
Comment (tianweidut) [reply client/starwhale/api/_impl/model.py:239]: ok

File: example/speech_command/sc/evaluator.py (hunk 1 above)
Comment (tianweidut) [reply example/speech_command/sc/evaluator.py:68]: the starwhale SDK did not define the precise ppl argument type requirements, so the users can customise by themselves.
https://github.com/star-whale/starwhale/blob/main/client/starwhale/api/_impl/model.py#L180

Comment (goldenxinxing) [reply example/speech_command/sc/evaluator.py:68]: ‘If the users don't config batch size, the example code keeps the original structure?’ yes, if don't config batch size there is no need to change the code, I just changed this demo to the implementation of the default batch

Comment (goldenxinxing) [reply example/speech_command/sc/evaluator.py]: done

File: example/mnist/mnist/evaluator.py (hunk 2 above)
Comment (goldenxinxing) [reply example/mnist/mnist/evaluator.py]: done

File: example/mnist/mnist/custom_evaluator.py (hunk 3 above)
Comment (goldenxinxing) [reply example/mnist/mnist/custom_evaluator.py:174]: this has been rollbacked to single

**Review by tianweidut — APPROVED** [review]
LGTM

**Pull request merged**